fix(pos): include Product Bundle components in reserved qty to preven…

LewisMojica opened the pull request:
## POS: Fix real-time stock for Product Bundle components to prevent overselling
Fixes: #49021

### Summary
When selling Product Bundles in POS, the available stock for the component items wasn’t reflected in real time. This allowed POS sessions to oversell items used in bundles and then hit negative stock on closing.

This PR centralizes POS “reserved” quantity logic and includes bundles in the reservation.

Prevents double-counting of reservations when computing how many bundles can be sold.

## Root cause
`get_pos_reserved_qty` only accounted for direct item lines in submitted POS Invoices. It ignored quantities reserved through Product Bundles.

In `get_bundle_availability` we subtracted pos_reserved_qty again at the item level, which led to underestimation (double subtraction) and inconsistent availability between items and bundles.


### Broke down the logic into two smaller functions

`get_direct_pos_reserved_qty(item_code, warehouse)`: existing behavior (direct item lines in submitted POS Invoices).

`get_bundle_pos_reserved_qty(item_code, warehouse)`: new aggregation that multiplies POS Invoice Item.stock_qty (for the bundle) by each Product Bundle Item.qty to compute the reserved quantity of component items.

get_pos_reserved_qty now returns direct_reserved + bundle_reserved.

### Before

POS shows component items as available even after multiple bundle sales.

Bundle availability was undertesting due to double subtraction, causing giving an error when trying to sell bundles even is enough components items where available.

Closing POS could fail with negative stock.

### After

Component reservations from submitted POS Invoices (both direct and via bundles) are reflected.

Bundle availability is computed without double subtracting reservations.

Eliminates overselling scenarios reproduced in #49021.

### Notes & Compatibility:
No schema changes.

Doesn’t alter valuation; only the reservation/availability logic seen by POS.

### Testing
Create items A (qty 5) and B (qty 5).

Create bundle “Combo”: 1×A, 2×B.

Open POS.

Submit a POS Invoice for 1×Combo.

Before (bug): Session 2 still shows A:5, B:5 and allows selling beyond component stock.

After (fix): Session 2 reflects A and B reserved via the bundle sale; attempting to oversell is blocked. Bundle availability matches component constraints.



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- Improvements
  - POS availability now distinguishes direct-sale reservations from bundle-component reservations; per-item gating uses raw bin stock for clearer limits.
  - Bundle caps refined so checkout limits reflect component stock after accounting for bundle reservations.
  - Packing lists are updated during validation to keep checkout data current.

- Bug Fixes
  - Aggregated reservation checks reduce overselling and improve real-time availability.

- Documentation
  - Added descriptive docstrings for the new reservation aggregation logic.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

LewisMojica committed:
fix(pos): include Product Bundle components in reserved qty to prevent overselling

- Add `get_bundle_pos_reserved_qty` to account for component items in submitted POS Invoices
- Update `get_pos_reserved_qty` to sum direct and bundle reservations
- Remove double subtraction in `get_bundle_availability` to avoid underestimating bundle availability
- Prevents overselling when multiple POS sessions sell bundles with shared components
- Fixes #49021

LewisMojica committed:
chore: remove unused variable

diptanilsaha commented on erpnext/accounts/doctype/pos_invoice/pos_invoice.py:
```diff
@@ -908,9 +924,33 @@ def get_pos_reserved_qty(item_code, warehouse):
-
 	return flt(reserved_qty[0].stock_qty) if reserved_qty else 0
 
+def get_bundle_pos_reserved_qty(item_code, warehouse):
```
Currently, the POS Invoice packed list is not used for Product Bundles or Repacked Items. I have enabled it on my PR by calling the function `self.update_packing_list()` inside the `validate` function, which will make use of the `packed_list` child table inside the POS Invoice. This will help you simplify the query for calculating the packed items as well.

You can refer to #49106

LewisMojica replied:
I implemented update_packed_items(), which was another small bug, thanks for pointing it out.

LewisMojica replied:
I think the approach you used is better.

I tested the behavior and confirmed that if a Product Bundle’s component quantities are modified after the sale, my current approach recalculates using whatever the current definition of the Product Bundle is, which makes historical reservations incorrect.

Given that, packed_items is clearly the more reliable source since it’s a snapshot of the bundle at the time of sale. Initially I wasn't sure due to concerns about relying on update_packed_items() being consistently correct (I didn’t want a hidden or future bug in that step to affect the accuracy of  
reservations). 

Also I'm a little bit concerned about this escenario:

1. An open POS sells a bundle
2. Update with new code happens while a POS session is still open.
3. That open session has already sold bundles (so there are unconsolidated POS Invoices with 
the packed_items table empty).

In that case the reserved qty will be under estimated till POSs are closed. after that all invoices with bundle will have the packed items populated. 

Maybe I'm too paranoid.

That said, the risk of incorrect data from recomputing off the live bundle definition outweighs the risk of a missing packed_items row.

LewisMojica replied:
changes applied in commit [d77d796](https://github.com/frappe/erpnext/pull/49108/commits/d77d79e01173d1831b67e21e3850fc931e834f4d)

LewisMojica committed:
chore: apply pre-commit formatting

LewisMojica committed:
fix(pos): populate packed_items table in POS Invoice

LewisMojica commented on erpnext/accounts/doctype/pos_invoice/pos_invoice.py:
```diff
@@ -868,9 +868,8 @@ def get_bundle_availability(bundle_item_code, warehouse):
-		available_qty = item_bin_qty - item_pos_reserved_qty
 
-		max_available_bundles = available_qty / item.qty
+		max_available_bundles = item_bin_qty / item.qty
```
Valid issue. In the product bundle form is possible to add item with qty 0 and cause an error at the pos (division by 0).
I'll fix it with max_available_bundles = item_bin_qty / item.qty if item.qty else 1000000
that way if a packed item qty is 0 it doesn't cause an error and don't cause max_avaible_bundle go to 0 when it shouldn't.

LewisMojica replied:
Actually I think this is a separate bug not related to #49021. Should be a separate PR.

LewisMojica replied:
sure

LewisMojica commented on erpnext/accounts/doctype/pos_invoice/pos_invoice.py:
```diff
@@ -908,9 +925,33 @@ def get_pos_reserved_qty(item_code, warehouse):
-
 	return flt(reserved_qty[0].stock_qty) if reserved_qty else 0
 
+def get_bundle_pos_reserved_qty(item_code, warehouse):
+	"""Reserved qty for the item as a component of Product Bundles in submitted POS Invoices (matching warehouse)."""
+
+	p_inv = frappe.qb.DocType("POS Invoice")
+	p_item = frappe.qb.DocType("POS Invoice Item")
+	pb = frappe.qb.DocType("Product Bundle")
+	pb_item = frappe.qb.DocType("Product Bundle Item")
+
+	bundle_reserved = (
+		frappe.qb.from_(p_inv)
+		.from_(p_item)
+		.from_(pb)
+		.from_(pb_item)
+		.select(Sum(p_item.stock_qty * pb_item.qty).as_("stock_qty"))
+		.where(
+			(p_inv.name == p_item.parent)
+			& (IfNull(p_inv.consolidated_invoice, "") == "")
+			& (p_item.docstatus == 1)
+			& (p_item.warehouse == warehouse)
+			& (pb.name == p_item.item_code)      # POS item is a bundle
+			& (pb_item.parent == pb.name)        # Bundle items
+			& (pb_item.item_code == item_code)   # This specific item
+		)
+	).run(as_dict=True)
+	return flt(bundle_reserved[0].stock_qty) if bundle_reserved else 0
```
Don't think this make sense because of this escenario:

1. Product bundle is sold from POS.
2. Some Bundle is set to disabled.
3. Now the available qty of items sold in the bundle will not be accurate as the one sold in the disabled bundle won't be computed.

LewisMojica replied:
You already handle disabled=0 where it matters: It won't be available to sell once it's disabled. Reservations should reflect what actually happened, regardless of current status.

LewisMojica committed:
fix: remove unclear message related to availability of product bundle

LewisMojica committed:
fix(pos): use packed_items snapshot for bundle reservations

Replaced live Product Bundle queries with `Packed Item` table
lookups to ensure historical reservation accuracy.
Addresses bundle qty underestimation and avoids errors when
bundle definitions change after sale.

Inspired by approach from @diptanilsaha in #49106.

diptanilsaha commented on erpnext/accounts/doctype/pos_invoice/pos_invoice.py:
```diff
@@ -893,13 +892,50 @@ def get_bin_qty(item_code, warehouse):
+	  warehouse (str): The Warehouse to filter by.
+
+	Returns:
+	  float: The total reserved quantity for the item in the given
```
`get_pos_reserved_qty_from_table` is going to return a float value, which is the total reserved quantity for the item in the warehouse for child_table from submitted, unconsolidated POS Invoices.

diptanilsaha commented on erpnext/accounts/doctype/pos_invoice/pos_invoice.py:
```diff
@@ -893,13 +892,50 @@ def get_bin_qty(item_code, warehouse):
+	qty_column = "qty" if child_table == "Packed Item" else "stock_qty"
 
-	reserved_qty = (
+	stock_qty = (
```
Rename it to `reserved_qty` as it sounds more clarifying than `stock_qty` in this case.

diptanilsaha commented on erpnext/accounts/doctype/pos_invoice/pos_invoice.py:
```diff
@@ -909,7 +945,7 @@ def get_pos_reserved_qty(item_code, warehouse):
 	).run(as_dict=True)
 
-	return flt(reserved_qty[0].stock_qty) if reserved_qty else 0
+	return stock_qty
```
return the float value instead of a list, i.e., replace `return stock_qty` with `return flt(reserved_qty[0].stock_qty) if reserved_qty else 0`

diptanilsaha commented on erpnext/accounts/doctype/pos_invoice/pos_invoice.py:
```diff
@@ -893,13 +892,50 @@ def get_bin_qty(item_code, warehouse):
+	pinv_item_reserved_qty = get_pos_reserved_qty_from_table("POS Invoice Item", item_code, warehouse)
+	packed_item_reserved_qty = get_pos_reserved_qty_from_table("Packed Item", item_code, warehouse)
+
+	reserved_qty = flt(pinv_item_reserved_qty[0].stock_qty) if pinv_item_reserved_qty else 0
```
Adding the values will be fine. Refer to https://github.com/frappe/erpnext/pull/49108#discussion_r2274198946

LewisMojica committed:
chore: improve code clarity per reviewer feedback

- Rename stock_qty variable to reserved_qty for clarity
- Update get_pos_reserved_qty_from_table to return float
- Simplify aggregation logic in get_pos_reserved_qty
- Ensure return values match docstring specifications
